### PR TITLE
Added nginx-proxy container and introducing .env

### PR DESCRIPTION
**Pull request by bojanbogdanovic**
Currently there is no HTTP reverse proxy tooling available in the OS developer repo. The goal here was to make OS projects easily setup via a few commands and configurations. I have looked into traefik but gave nginx-proxy the preference, due it's simple use and available advanced configurations that Nginx offers.

### CODE REVIEW

**Comment by navneet0693** [comment]
@bojanbogdanovic Nice one! Long time pending. Please also update: https://github.com/goalgorilla/drupal_social/blob/master/drush/drush.yml#L10

**Review by Kingdutch — CHANGES_REQUESTED** [review]
Hey Bojan,

While I used Traefik for the Tugboat setup and like how it works, this PR would break my local development workflow.

I currently use `nginxproxy`, which uses the `VIRTUAL_HOST` environment variables to make environments available under a specific URL.

The PR introduces two problems for me specifically:
- It binds a container to port 80 on the host which disallows running multiple Open Social installations at the same time (I've previously done this when comparing enterprise sites and I've recently done it for extensions development and/or to compare a site with extension to one without the extension).
- It removes all the `VIRTUAL_HOST` annotations, breaking usage with `nginxproxy`

A compromise would be to leave the environment variables for `nginxproxy` in place and add the traefik container, with configuration, without binding that container to port 80 (people could do that in a docker.override.yml file). (This assumes that the traefik config on the containers does not break them responding to requests through nginxproxy).

Alternatively I'm happy to switch from nginxproxy to Traefik (since I see I can still use `social.localhost` with it) but would want to retain the ability to run multiple Open Social installations next to each other :D

**Comment by Kingdutch** [comment]
In addition to the above, please use this moment to standardise on `.localhost` rather than `.local` (the `drush.yml` that Navneet referred to is correct in this case and does not need changing).

[`.localhost`](https://en.wikipedia.org/wiki/.localhost) is reserved by the ETF and will never be a valid extension. While [`.local`](https://en.wikipedia.org/wiki/.local) is also reserved as a special TLD it is used in local networking and multicast services, it is not reserved to loop-back to the local machine (which is what we want). So configuring `.local` domains to point to your local machine may break applications in fun and interesting ways in the future.

**Comment by bojanbogdanovic** [comment]
@Kingdutch thanks for the feedback. I have updated the PR with the same approach as "Docker 4 Drupal": they have a `.env` available where they can set the project port. The downside is that we now have to access the URL with the port number: `social.localhost:8000`.

**Comment by Kingdutch** [comment]
I think adding a `.env` is a good idea :D We may want to `.gitignore` the `.env` file and distribute a `.env.dist` file. That way people can modify `.env` without merge conflicts and update the repository around it.

The downside of not being able to bind to port `80` will always be there as long as the router (i.e. `traefik` or `nginxproxy`) is included in the docker-compose file rather than separately.

For nginx I had created the following aliases to manage it:
```bash
# Management for the proxy to our docker containers so we can use hostnames instead of ports
alias nginxproxy='docker run -d -p 80:80 -v ~/.nginxproxy/gateway-timeout.conf:/etc/nginx/conf.d/gateway-timeout.conf:ro -v /var/run/docker.sock:/tmp/docker.sock:ro nginxproxy/nginx-proxy'
alias nginxproxy-name="docker ps --format '{{.Image}}\\t{{.Names}}' | awk 'BEGIN { FS=\"\t\" } \$1 == \"nginxproxy/nginx-proxy\" { print \$2 }'"
alias nginxproxy-connect="nginxproxy-name | xargs docker network connect"
alias nginxproxy-logs="nginxproxy-name | xargs docker logs -f"
alias nginxproxy-stop="nginxproxy-name | xargs docker stop"
```

Starting my computer fresh and opening a project would mean I'd:
```bash
nginxproxy # Start the nginxproxy container
nginxproxy-connect drupal_social_default # Connect to the network for my drupal_social folder
docker-compose up # start the drupal_social folder
```

Any other project could then be added with
```bash
nginxproxy-connect <network>
docker-compose up # within the folder of that project
```

We could do a similar thing with `traefik` where we instruct people on how to start up Traefik outside of their project and then connect it to the project. That way they'll only have a single router running that makes all their docker projects available on port 80.

I'm happy to experiment with Traefik over nginxproxy since the latter does have some connection timeouts in some edge cases (specifically some SAML redirects or large pages after a redirect). I do think we should make it documentation and not part of the `docker-compose.yml` though (or maybe included as `docker-compose.router.yml`? so that it's still there for people to easily start up from within a single project if needed).

**Review by Kingdutch — APPROVED** [review]

File: nginx/custom-nginx.conf [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+proxy_read_timeout          600;
+send_timeout                600;
+
+# Also increase (header) buffer size
+proxy_buffer_size   128k;
+proxy_buffers   4 256k;
+proxy_busy_buffers_size   256k;
+
+# Fix issues with large json responses
+proxy_max_temp_file_size 0;
```
Comment (Ressinel) [inline nginx/custom-nginx.conf:12]: No newline at end of file
```suggestion
proxy_max_temp_file_size 0;

```

**Comment by ronaldtebrake** [comment]
LGTM!

Would be good to maybe write something in the PR description, or in slite and link that to the PR. 
Especially for historical purposes for those in the future wondering why we went this route.
But for me that could be done in paralel so we can start working with this awesomeness :)

**Pull request merged**